File: 2545-height-of-binary-tree-after-subtree-removal-queries/2545-height-of-binary-tree-after-subtree-removal-queries.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def treeQueries(self, root: Optional[TreeNode], queries: List[int]) -> List[int]:

        def recur(nn):

            if nn.left:
                ll, la = recur(nn.left)
                if nn.right:
                    rl, ra = recur(nn.right)
                else:
                    la[ll[-1]] = 1  # if ll[-1] is deleted, max hight = 0
                    ll.append(nn.val)
                    return ll, la
            elif nn.right:
                rl, ra = recur(nn.right)
                ra[rl[-1]] = 1  # if rl[-1] is deleted, max hight = 0
                rl.append(nn.val)
                return rl, ra
            else:
                return [nn.val], {}

            if len(ll) >= len(rl):
                la[ll[-1]] = len(rl) + 1
                ll.append(nn.val)
                return ll, la
            else:
                ra[rl[-1]] = len(ll) + 1
                rl.append(nn.val)
                return rl, ra

        longest, alt_d = recur(root)
        ll = len(longest) - 1
        new_alt = 0
        new_alt_d = {}
        for i, x in enumerate(longest[::-1][1:]):
            alt = alt_d[x] + i - 1
            new_alt = max(new_alt, alt)
            new_alt_d[x] = new_alt

        return [ new_alt_d[q] if q in new_alt_d else ll for q in queries ]
        

      

File: 2545-height-of-binary-tree-after-subtree-removal-queries/2545-height-of-binary-tree-after-subtree-removal-queries.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int height[100001];
    int level[100001];
    int levelMaxHt[100001];
    int levelSecondMax[100001];

    int findHeight(TreeNode* root,int lvl){
        if(root == NULL) return 0;
        level[root->val] = lvl;
        int left = findHeight(root->left,lvl+1);
        int right = findHeight(root->right,lvl+1);
        height[root->val] = 1 + max(left,right);

        if(levelMaxHt[lvl] < height[root->val]){
            levelSecondMax[lvl] = levelMaxHt[lvl];
            levelMaxHt[lvl] = height[root->val];
        }
        else if(levelSecondMax[lvl] < height[root->val]){
            levelSecondMax[lvl] = height[root->val];
        }
        return height[root->val];
    }
    vector<int> treeQueries(TreeNode* root, vector<int>& queries) {
        findHeight(root,0);
        vector<int> result;
        for(auto& node : queries){
            int l = level[node];
            int level_max_ht = levelMaxHt[l];
            int h = height[node];
            int ans = l +(level_max_ht == h ? levelSecondMax[l] : levelMaxHt[l]) - 1;
            result.push_back(ans);
        }
        return result;
    }
};
/*
1. jis node ko delete krna h vo kis level pr h or uski kitni height h aur uss level ki max height aur second max height kya h.
2. agar jis node ko delete kr rahe ho aur ussi ki uss level pr max height h toh second max height nikalo .
3. formula = (node jis level pr h) + (height of other nodes on that level) -1

har ek node ki height aur level chahiye toh precompute it.and har levelki max height and second max height 
*/